### ISSUE TITLE
GRPO训练，用colocate模式在rollout阶段会出现模型参数或者优化器参数没有被offload导致oom的问题

### ISSUE THREAD

**Issue by 8125345** [modelscope/ms-swift#9991]
### Checklist / 检查清单

- [x] I have searched existing issues, and this is a new question or discussion topic. / 我已经搜索过现有的 issues，确认这是一个新的问题与讨论。

### Question Description / 问题描述

问题现象：
用GRPO训练qwen3.5-35B模型，用colocate模式，在训练过程中会出现模型或者优化器参数没有被offload从而导致OOM的问题。offload_optimizer和offload_model都设置成了true了，但是在中间的几个step有时候就会出现某张卡的显存一直没有被释放的现象，导致再次wake up weight就会oom。
分析原因是因为数据太长了，导致vllm在rollout阶段，还没有完全推理完，一直占用显存，所以将
--max_length 8192 
--max_completion_length 2048 设置成这样就能解决这个问题；但是新的问题出现了，在训练完模型推理的时候发现很多回答被截断了，所以怀疑是因为训练将max_completion_length长度减少的原因。（或者是其他原因？）
所以想问下，有没有一种方法我可以设置max_completion_length的大一些，同时也不会影响GRPO训练过程的oom，想到的方法是：能不能等每次rollout阶段vllm生成completions完全生成完后再wake up 权重，这样就不会出现oom的情况了。

**Comment by hjh0119** [comment]
> 能不能等每次rollout阶段vllm生成completions完全生成完后再wake up 权重

现在就是这么做的

**Comment by 8125345** [comment]
> > 能不能等每次rollout阶段vllm生成completions完全生成完后再wake up 权重
> 
> 现在就是这么做的
@hjh0119 但是在训练过程中，第二个节点的倒数两张卡总会占用很高的显存，跟其他卡的显存完全不一致。

<img width="989" height="746" alt="Image" src="https://github.com/user-attachments/assets/605cb9e2-1165-44c9-a3fd-8f88d99e32a6" />
这是我的启动脚本：
nnodes=2
nproc_per_node=8

NNODES=$nnodes \
NODE_RANK=0 \
MASTER_ADDR=10. \
MASTER_PORT=29500 \
PYTORCH_CUDA_ALLOC_CONF='expandable_segments:True' \
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
NPROC_PER_NODE=$nproc_per_node \
IMAGE_MAX_TOKEN_NUM=512 \
megatron rlhf \
    --rlhf_type grpo \
    --model '/modelfile/Qwen3.5-35B-A3B/' \
    --dataset '' \
    --external_plugins /plugin_custom.py \
    --reward_funcs llm_score \
    --output_dir '/测试用/' \
    --num_generations 8 \
    --micro_batch_size 1 \
    --global_batch_size 16 \
    --steps_per_generation 1 \
    --context_parallel_size 1 \
    --tensor_model_parallel_size 1 \
    --pipeline_model_parallel_size 8 \
    --expert_model_parallel_size 2 \
    --finetune true \
    --agent_template 'qwen3_5' \
    --tuner_type full \
    --freeze_vit true \
    --freeze_aligner false \
    --torch_dtype bfloat16 \
    --loss_type grpo \
    --lr 1e-6 \
    --bf16 true \
    --beta 0.04 \
    --epsilon 0.2 \
    --epsilon_high 0.28 \
    --importance_sampling_level sequence \
    --use_vllm true \
    --vllm_mode colocate \
    --vllm_gpu_memory_utilization 0.4 \
    --vllm_tensor_parallel_size 4 \
    --sleep_level 2 \
    --offload_model true \
    --offload_bridge true \
    --offload_optimizer true \
    --optimizer_cpu_offload true \
    --use_precision_aware_optimizer true \
    --temperature 0.9 \
    --repetition_penalty 1.1 \
    --vllm_mm_processor_cache_gb 0 \
    --enable_thinking false \
    --save_safetensors true \
    --moe_permute_fusion true \
    --moe_grouped_gemm true \
    --moe_shared_expert_overlap true \
    --moe_aux_loss_coeff 1e-6 \
    --recompute_granularity full \
    --recompute_method uniform \
    --recompute_num_layers 1 \
    --dataset_num_proc 64 \
    --dataloader_num_workers 8 \
    --overlong_filter true \
    --padding_free false \
    --sequence_parallel true \
    --dynamic_sample false \
    --attention_backend flash \
    --num_train_epochs 1 \
    --truncation_strategy left \
    --max_length 8192 \
    --max_completion_length 2048 \
    --save_steps 5 \
    --save_total_limit 2 \
    --no_save_optim true \
    --no_save_rng true \
    --metric_for_best_model reward \
    --logging_steps 1 \
    --log_completions true \

**Comment by hjh0119** [comment]
感觉像对应的卡还在rollout，rollout完会卸载

**Comment by 8125345** [comment]
> 感觉像对应的卡还在rollout，rollout完会卸载

@hjh0119 但是每次都是这两张卡一直高显存占用（可能是因为rollout），导致下次wake up 权重开始训练的时候这两张卡就会OOM报错退出

**Comment by 8125345** [comment]
> 感觉像对应的卡还在rollout，rollout完会卸载

@hjh0119 还有一个信息是，如果用很短的数据去训练，就能训起来，最后那两张卡的显存占用也会很低，但一旦用很长的数据，比如超过5000个中文字符的数据去训，就会出现上面的情况，但是我这边的场景确实需要很长的输入去训练，输出也比较长。

**Comment by 8125345** [comment]
现在换成vllm serve模式，也就是rollout单独占用了一台机器，剩下两台只做训练，发现还是有相同的问题，最后一个节点的最后两张卡占用显存异常

<img width="990" height="549" alt="Image" src="https://github.com/user-attachments/assets/368deeb6-7f44-4c1b-bb0e-5355bf67d209" />
不确定是什么问题，GRPO训练的过程中，最后两张卡是会做什么运算吗？我这里--beta 0.04，也就是会用到ref模型，是不是最后两张卡在加载ref模型？

**Comment by hjh0119** [comment]
PP last rank 会有额外的开销（lm_head)，显存偏高是正常的，可以设置 decoder_first_pipeline_num_layers 参数来平衡
